### DISCUSSION TITLE
Welcome to TornadoVM Discussions!

### DISCUSSION THREAD

**Discussion by jjfumero** [beehive-lab/TornadoVM#45] [announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by adelahmed** [comment]
I'm trying to install TornadoVM on my machine. My machine configuration:
- Dell Laptop
- Intel i7 - 11 Gen
- Intel Iris Xe Graphics
- Nvidia GeForce MX350
- Windows 11

I'm following the steps listed on [TornadoVM Documentation Page](https://tornadovm.readthedocs.io/en/latest/installation.html#tornadovm-for-windows-10-11-using-graalvm).

I've installed the prerequisites, downloaded TornadoVM, GraalVM, installed CUDA SDK. But when I try to compile TornadoVM I get the following error:

`
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.11.0:compile (default-compile) on project tornado-drivers-common: Compilation failure: Compilation failure:

[ERROR] C:\dev\TornadoVM\tornado-drivers\drivers-common\src\main\java\uk\ac\manchester\tornado\drivers\common\compiler\phases\guards\ExceptionSuppression.java:[63,22] error: cannot find symbol
[ERROR]   symbol:   method getAnchoredNode()
[ERROR]   location: variable anchor of type ValueAnchorNode
[ERROR] C:\dev\TornadoVM\tornado-drivers\drivers-common\src\main\java\uk\ac\manchester\tornado\drivers\common\compiler\phases\guards\ExceptionSuppression.java:[66,22] error: cannot find symbol
[ERROR]   symbol:   method removeAnchoredNode()
[ERROR]   location: variable anchor of type ValueAnchorNode
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :tornado-drivers-common
Error running 'mvn clean': Command '['mvn', '-T1.5C', '-Dcmake.root.dir=$CMAKE_ROOT', '-Pgraal-jdk-21,ptx-backend,opencl-backend', 'install']' returned non-zero exit status 1.
Traceback (most recent call last):
  File "/c/dev/TornadoVM/bin/compile", line 373, in <module>
    main()
  File "/c/dev/TornadoVM/bin/compile", line 365, in main
    mvn_build_result = build_tornadovm(args, backend_profiles)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/c/dev/TornadoVM/bin/compile", line 264, in build_tornadovm
    return result
           ^^^^^^
UnboundLocalError: cannot access local variable 'result' where it is not associated with a value
make: *** [Makefile:15: graal-jdk-21] Error 1
`

It seems like the compilation is looking for the wrong version of the ValueAnchorNode class in GraalVM library jars. Any help on how to overcome this error?

The command I issue to compile TornadoVM is:
`make graal-jdk-21 BACKEND=ptx,opencl`

**Reply by adelahmed** [reply]
### OK. Issue solved.

I followed the tutorial on the web page, which instructed me to download the following GraalVM JDK 21

https://download.oracle.com/graalvm/21/latest/graalvm-jdk-21_windows-x64_bin.zip

which didn't work.

The error was complaining that some method was missing from a class. This usually happens when new methods are added in more recent revisions of the code. So this prompted me to test with a more recent release of the GraalVM JDLK

I went to the GraalVM JDK archive site:

https://www.oracle.com/java/technologies/javase/graalvm-jdk21-archive-downloads.html

and downloaded GraalVM JDK **21.0.1** from here:

https://download.oracle.com/graalvm/21/archive/graalvm-jdk-21.0.1_windows-x64_bin.zip

then updated my `sources.env` to point to the new GraalVM JDK path, fired up the compilation command and it worked.

I hope this small detail is mentioned somewhere on the documentation webpage so that others don't go through what I did.

**Reply by jjfumero** [reply]
Hi @adelahmed . Thank you for the report. We will check and update the documentation accordingly.

**Comment by pezi** [comment]
Tested tornado with `make tests` on  an Apple M2

I am unsure, is this kind of failure OK?

```
[TornadoVM] Warning: The loop bounds will be configured by the GridScheduler. Check the grid by using the flag --threadInfo.
Test: class uk.ac.manchester.tornado.unittests.compute.ComputeTests
	Running test: testNBodyBigNoWorker       ................  [FP64 UNSUPPORTED FOR CURRENT DEVICE] 
	Running test: testDFTVectorTypes         ................  [PASS] 
	Running test: testBlackScholes           ................  [PASS] 
	Running test: testRenderTrack            ................  [PASS] 
	Running test: matrixVectorFloat4         ................  [PASS] 
	Running test: testNBody                  ................  [FP64 UNSUPPORTED FOR CURRENT DEVICE] 
	Running test: testNBodySmall             ................  [FP64 UNSUPPORTED FOR CURRENT DEVICE] 
	Running test: testDFTDouble              ................  [FP64 UNSUPPORTED FOR CURRENT DEVICE] 
	Running test: testDFTFloat               ................  [PASS] 
	Running test: testHilbert                ................  [PASS] 
	Running test: testMontecarlo             ................  [FP64 UNSUPPORTED FOR CURRENT DEVICE] 
	Running test: testMandelbrot             ................  [FAILED] 
		\_[REASON] expected:<8> but was:<9>
	Running test: testEuler                  ................  [PASS] 
	Running test: testJuliaSets              ................  [FAILED] 
		\_[REASON] expected:<-1000.0> but was:<2.0576131>
	Running test: matrixVector               ................  [PASS] 
Test ran: 15, Failed: 2, Unsupported: 5
```

```
tornado --version                                                               
version=1.0.2-dev
branch=master
commit=c36582a

Backends installed: 
	 - opencl
```

```
./bin/tornadovm-installer --jdk graalvm-jdk-21  --backend opencl
```

**Reply by jjfumero** [reply]
Hi @pezi , yes, we can reproduce these errors as well. It looks to me related to FP32 precision errors. We have checked the image for both fractals (Mandelbrot and Julia) and they look correct though.
